test/util/framework/identities_helper.go

```diff
@@ -0,0 +1,660 @@
+	}
+
+	if err := state.initializeUnlocked(leasedRGs); err != nil {
+		return fmt.Errorf("failed to initialize managed identities pool state: %w", err)
+	}
+
+	return nil
+}
+
+func (tc *perItOrDescribeTestContext) AssignIdentityContainers(ctx context.Context, count uint8, waitBetweenRetries time.Duration) error {
+	state, err := NewLeasedIdentityPoolState(tc.msiPoolStateFilePath())
+	if err != nil {
+		return fmt.Errorf("failed to open managed identities pool state file: %w", err)
+	}
+	defer state.Close()
+
+	for {
+		err := state.AssignNTo(specID(), count)
+		if err == nil {
+			return nil
+		}
+		if !errors.Is(err, ErrNotEnoughFreeIdentityContainers) {
+			return err
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(waitBetweenRetries):
+		}
+	}
+}
+
+func (tc *perItOrDescribeTestContext) GetLeasedIdentities() (LeasedIdentityPool, error) {
+
+	state, err := NewLeasedIdentityPoolState(tc.msiPoolStateFilePath())
+	if err != nil {
+		return LeasedIdentityPool{}, fmt.Errorf("failed to open managed identities pool state file: %w", err)
+	}
+	defer state.Close()
+
+	leasedRG, err := state.UseNextAssigned(specID(), tc.registerLeasedIdentityContainer)
+	if err != nil {
+		return LeasedIdentityPool{}, fmt.Errorf("failed to lease next managed identities resource group: %w", err)
+	}
+
+	return LeasedIdentityPool{
+		ResourceGroupName: leasedRG,
+		Identities:        NewDefaultIdentities(),
+	}, nil
+}
+
+func (tc *perItOrDescribeTestContext) ReleaseLeasedIdentities(ctx context.Context, leasedContainers []string) error {
+
+	if !tc.UsePooledIdentities() || len(leasedContainers) == 0 {
+		return nil
+	}
+
+	creds, err := tc.perBinaryInvocationTestContext.getAzureCredentials()
+	if err != nil {
+		return err
+	}
+	subscriptionID, err := tc.getSubscriptionIDUnlocked(ctx)
+	if err != nil {
+		return err
+	}
+	client, err := armauthorization.NewRoleAssignmentsClient(subscriptionID, creds, nil)
+	if err != nil {
+		return err
+	}
+
+	state, err := NewLeasedIdentityPoolState(tc.msiPoolStateFilePath())
+	if err != nil {
+		return fmt.Errorf("failed to open managed identities pool state file: %w", err)
+	}
+	defer state.Close()
+
+	var errs []error
+	for _, resourceGroup := range leasedContainers {
+		err := state.ReleaseByContainerName(resourceGroup, func() error {
+			return tc.cleanupLeasedIdentityContainer(ctx, client, resourceGroup)
+		})
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to release identity container %s: %w", resourceGroup, err))
+		}
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("failed cleanup operations: %w", errors.Join(errs...))
+	}
+	return nil
+}
+
+func (tc *perItOrDescribeTestContext) msiPoolStateFilePath() string {
+	return filepath.Join(tc.perBinaryInvocationTestContext.sharedDir, msiPoolStateFileName)
+}
+
+func (tc *perItOrDescribeTestContext) registerLeasedIdentityContainer(resourceGroupName string) error {
+	tc.contextLock.Lock()
```

Comment:
Why is this state disjoint from the on-disk state? Prefer one source of truth and one lock

Reply:
Cleanup now works off the state file without an intermediate state